### ISSUE TITLE
App 多页应用页面跳转无法接收 url 参数.

### ISSUE THREAD

**Issue by acrazing** [baidu/amis#7109]
#### 描述问题：

多页应用 button 跳转 url 带参数无法解析.

#### 截图或视频：

这是直接打开页面的效果, ID 能拿到.
![image](https://github.com/baidu/amis/assets/10029810/6743f339-051d-4676-aa3d-5547744e24af)

这是从另一个页面跳转过来的效果, ID 拿不到.
![image](https://github.com/baidu/amis/assets/10029810/e199ea7e-c82f-4711-b935-4cc67dfccb09)

#### 如何复现（请务必完整填写下面内容）：

1. 你是如何使用 amis 的？

    `npm + vite`

3. amis 版本是什么？请先在最新 beta 版本测试问题是否存在

    `3.1.1`, 已经是最新版了

5. 粘贴有问题的完整 `amis schema` 代码：

    在线体验地址: <https://codesandbox.io/s/amis-app-redirect-bug-demo-9ii85t?file=/src/main.tsx>.

7. 操作步骤

    在 From 页面, 点击按钮跳转到 To 页面, 拿不到 ID.

**Comment by 2betop** [comment]
如果外面是单页应用，需要下发最新的 location 对象

**Comment by acrazing** [comment]
我研究了一下这个 Bug, 发现似乎是因为 `watchRouteChange` 仅作用于 `App` 组件, 而其它组件依赖于 `location` 属性的更新来判断是否要重新解析 query 参数. 然后我就尝试每次 location 变更都传一个新 location 给 render (history 的 location 是不可变的), 但是又产生了个新问题: 如果是跳转页面是 CRUD 并且设置了 syncLocation 的话, 它会加载两次, 因为它总是会默认加个 `page=1` 参数到 url 上. 所以我实际代码最终给的魔法是:

```ts
function Root() {
  const [location, setLocation] = useState(location);
  useEffect(() => {
    let lastPathname = location.pathname;
    return history.listen(() => {
      if(lastPathname !== location.pathname) {
        lastPathname = location.pathname
        // 只有路径变化的时候才重新渲染
        setLocation(new Proxy(
          location, // 不能是 history.location, 因为参数变更时不能接收到新的
          { get: (t, p) => location[p] }
        ))
      }
    });
  }, []);
  return render(
    schema,
    {
      ...otherProps, // 这里不能写字面量, 比如我要传个 env, 会导致完全重新渲染
      location
    },
    {
      // ...
      // 不给 watchRouteChange, App 也有监听 location 属性变化
    }
  );
}
```

差点弃疗了..

**Comment by acrazing** [comment]
> 如果外面是单页应用，需要下发最新的 location 对象

@2betop 这样和 CRUD 有冲突, 会刷新两次, 你看我上面说的那个.

**Comment by 2betop** [comment]
对了 env.updateLocation 有两个参数，第二个是否用 replace，当 page=1 其实是应该用 history.replace 弄上去的，不知道你那边目前是怎么处理的

**Comment by acrazing** [comment]
是的, 而且我还加了如果跳转的 pathname 和当前一样, 也用 replace, 因为在发现这个问题前, 没有传新 location 给 render, 如果在 CRUD 里面换了筛选条件或者切页面, 完事后返回上一页页面不会刷新.

**Comment by acrazing** [comment]
我感觉这里有两个问题: 一个是 CRUD 和 App 处理 location 的逻辑有冲突, 另一个是 render('app') 的 location 更新机制不太好, 并且缺少文档.

**Comment by acrazing** [comment]
另外还有个优化的点: CRUD 可以加个参数, 判定在 page 或者筛选条件发生变更时是 `push` 还是 `replace`, 看起来目前没有这个参数支持?

**Comment by 2betop** [comment]
没有 cancel 可能的原因是自己实现了 fetcher 吧，这里需要支持这个逻辑 https://github.com/baidu/amis/blob/master/examples/embed.tsx#L182-L186

按理是不会请求两次，像我们的平台和 amis example 里面就没出现这个情况。能否提供一个复现你说这个情况的 repo？

**Comment by acrazing** [comment]
有的:

<img width="771" alt="image" src="https://github.com/baidu/amis/assets/10029810/f29bc2b0-1c87-4269-b7a5-21c27a9742ff">

另外我改造了下上面那个 demo: https://codesandbox.io/s/amis-app-redirect-bug-demo-9ii85t?file=/src/main.tsx

你可以体验一下, 或者看看应该怎么优化?
